Migrate Puppeteer tests to Playwright

swissspidy opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

As per https://make.wordpress.org/core/2022/03/23/migrating-wordpress-e2e-tests-to-playwright/, this migrates all Puppeteer usage in core to Playwright.

**To-do**

* [ ] ~Maybe set up ESLint/Prettier for formatting...? Odd that core doesn't have this yet 🤷~ see https://core.trac.wordpress.org/ticket/31823
* [ ] Publish update on make/core after merging
	- Mention the change from `tests/visual-regression/specs/__image_snapshots__` to `tests/visual-regression/specs/__snapshots__` for visual tests snapshots

Trac ticket: https://core.trac.wordpress.org/ticket/59517

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

swissspidy committed:
WIP Playwright migration

swissspidy committed:
Convert visual regression tests

swissspidy committed:
Fix mask config

swissspidy committed:
Initial conversion of edit posts tests

swissspidy committed:
Fix file ext

swissspidy committed:
Convert dashboard test

swissspidy committed:
Convert another test

swissspidy committed:
Improve test

swissspidy committed:
Remove types package again

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Remove debug cruft

swissspidy committed:
Install browsers separately

Mamaduka commented:
Fantastic work, @swissspidy! 🦸

I'll start going through the tests and change the code into an a11y locator where possible.

A general note regarding the Gutenberg plugin test: Maybe we should handle installation via `wp-env`. As far as I can tell, it doesn't need to test installation; just that activation doesn't cause fatal errors.

Cc-ing @kevin940726 and @WunderBart, in case they've any notes regarding the setup.

swissspidy commented:
> A general note regarding the Gutenberg plugin test: Maybe we should handle installation via `wp-env`. As far as I can tell, it doesn't need to test installation; just that activation doesn't cause fatal errors.

Core doesn't use wp-env. But yes, I don't like how the test is doing a fresh install. It takes a long time and is brittle and difficult to clean up if one of the steps fails. But that's just how it was before, I didn't change any of the logic.

What we could do is downloading Gutenberg separately in the CI step, but that would mean you can't easily run the test locally. So right now I don't see a good solution.

Mamaduka commented:
Sounds good. We can improve this in follow-up. It doesn't need to be a blocker here.

Mamaduka commented on tests/e2e/playwright.config.js:
```diff
@@ -0,0 +1,64 @@
+		video: 'on-first-retry',
+	},
+	webServer: {
+		command: 'npm run wp-env start',
```
I think this should be `npm run env:start`.

swissspidy replied:
Ah yes. Copy pasta. Good catch 👍

swissspidy committed:
Fix webServer command

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy commented:
Hmm getting `Error: Cannot find module '@playwright/test/cli'` errors on CI right now, coming from `node_modules/@wordpress/scripts/scripts/test-playwright.js` 🤔

swissspidy committed:
Some more bug fixes

Mamaduka committed:
Update dashboard tests

Mamaduka commented:
Locally or on CI?

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -1,29 +1,32 @@
+	} ) => {
+		await admin.visitAdminPage( '/edit.php' );
+		await expect( page.locator(
 			'//td[text()="No posts found."]'
```
This can be replaced with a role-based selector `getByRole( 'cell', { name: 'No posts found.' } )`

Mamaduka replied:
This is an odd one. Chrome a11y tree gives the element the role of `gridcell`, but PW sees it as a `cell`.

swissspidy replied:
Maybe a different browser version or something?

kevin940726 replied:
FYI Playwright doesn't use the chromium accessibility API for `getByRole`. They implement their own resolution logic (following the standard ofc) in JS instead to solve the cross-browser compatibility issue. So in short the devtools in chromium isn't always going to match what Playwright sees 😆.

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -1,29 +1,32 @@
+		await editor.publishPost();
+		await admin.visitAdminPage( '/edit.php' );
 
 		await page.waitForSelector( '#the-list .type-post' );
```
`$$` and `$x` selectors are discouraged, so let's use locators instead, which will also allow us to drop the use of `waitForSelector`.

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -40,19 +43,20 @@ describe( 'Edit Posts', () => {
 		expect( postTitle.length ).toBe( 1 );
```
We can also use locator assertions here, e.g. `expect(firstPost).toHaveText(title)`

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -69,22 +73,24 @@ describe( 'Edit Posts', () => {
 		// Tab to the Quick Edit button and press Enter to quick edit.
-		await pressKeyTimes( 'Tab', 2 );
+		await page.keyboard.press( 'Tab' );
+		await page.keyboard.press( 'Tab' );
```
If we're using the playwright e2e utils package here, we can use the `pressKeys` util from the `PageUtils` fixture:

```js
await pageUtils.pressKeys( 'Tab', { times: 2 } )
```

swissspidy replied:
Ah I was looking for that!

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -69,22 +73,24 @@ describe( 'Edit Posts', () => {
-			`//a[contains(@class, "row-title")][contains(text(), "${ title }")]`
-		);
-		await editLink.focus();
+		await page.locator( `//a[contains(@class, "row-title")][contains(text(), "${ title }")]` ).focus();
```
This can be replaced with

```js
await page.getByRole('link', { name: title }).focus();
```

Mamaduka commented:
@WunderBart, I'm working on best practices updates, so that test standards match what we have for Gutenberg.

swissspidy commented:
> Locally or on CI?

That was only on CI so far

WunderBart commented:
Thanks for working on this, @swissspidy!

I've started submitting some suggestions, but I'll stop now since I can see this is actively being addressed as I scroll through the files. The few suggestions I submitted can be treated more generally because they apply to other files as well - I just didn't want to copy-paste the same comment. Hope it's OK!

Mamaduka commented:
Thank you, @WunderBart!

Mamaduka committed:
Update edit post tests

Mamaduka committed:
Use keyPress util

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -53,24 +49,24 @@ test.describe( 'Edit Posts', () => {
 
 		// Click the post title (edit) link
-		await page.locator( `//a[contains(@class, "row-title")][contains(text(), "${ title }")]` ).click();
+		await listTable.getByRole( 'link', { name: `“${ title }” (Edit)` } ).click();
```
If we care about an exact match, we need to explicitly set [the `exact` option](https://playwright.dev/docs/api/class-framelocator#frame-locator-get-by-role-option-exact) to `true` as it's `false` by default. If we're OK with a partial match, the `name: title` should be sufficient here!

Mamaduka replied:
A few other quick action links in the table row contain titles, and the test needs to click on the edit link. With suffix, there's no need to exact match.

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -53,24 +49,24 @@ test.describe( 'Edit Posts', () => {
-		const editorFrame = await page.waitForSelector( 'iframe[name="editor-canvas"]' );
-
-		const innerFrame = await editorFrame.contentFrame();
+		await page
+				.frameLocator( '[name=editor-canvas]' )
+				.locator( 'body > *' )
+				.first()
+				.waitFor();
```
I know this is how we do it in Gutenberg, but I wonder if going down the extra `> *` level is necessary. Shouldn't we achieve the same result with the following?

```js
await page.frameLocator('[name=editor-canvas]').locator('body').waitFor();
```

(This is a workaround because the `frameLocator` API doesn't implement a `waitFor` method.)


cc: @kevin940726

Mamaduka replied:
I don't have a strong opinion on this one. I copied it from the createNewPost helper.

kevin940726 replied:
I recall in some weird cases the canvas renders an empty `<body>` but then remounts itself and renders a different document. I believe it's a workaround of how the editor frame works now. I don't know if the issue has been resolved though.

WunderBart replied:
Interesting, I'd assume that since it's lazily evaluated that wouldn't make a difference. Any idea where that issue occurred?

Mamaduka replied:
I'm guessing it's due to some internals of block editors `IFrame` component.

WunderBart commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -53,24 +49,24 @@ test.describe( 'Edit Posts', () => {
 
-		// Wait for title field to render onscreen.
-		await innerFrame.waitForSelector( '.editor-post-title__input' );
+		const editorPostTitle = editor.canvas.getByRole( 'textbox', { name: 'Add title' } );
```
This looks like an anti-pattern to me. Above, we explicitly wait for the `editor-canvas` frame to be ready via [`frameLocator()`](https://playwright.dev/docs/api/class-page#page-frame-locator), and here we reference canvas by the `canvas` object, [which is instantiated](https://github.com/WordPress/gutenberg/blob/537316ad5666c5574a4370c4f285e5f953315739/packages/e2e-test-utils-playwright/src/editor/index.ts#L40) by [`frame()`](https://playwright.dev/docs/api/class-page#page-frame).

Shouldn't the `canvas` object be instantiated by the `frameLocator` so we don't have to juggle like that? 

The only reason we need to add the `frameLocator...waitFor` step is that the `frame` locator does not support auto-waiting, so if the `iframe` element is not available at the moment of calling `frame`, the canvas will refer to the `page` object instead which will fail the test. 

cc: @kevin940726

Mamaduka replied:
Selecting block editor elements via `editor.canvas.locator` in Gutenberg tests is standard practice. I'm trying to keep these similar.

WunderBart replied:
Yeah, that part should stay as is, only that the `editor.canvas` should be calling `frameLocator` instead of `frame`. I'll try spinning up a PR with the refactor to make it clearer.

kevin940726 replied:
`frameLocator` for some reason doesn't allow some APIs like `evaluate`, which makes calling frame-level APIs a bit awkward. For instance, if we want to get the active element in the frame wme will then have to do `editor.canvas.locator('html').evaluate(() => document.activeElement.innerHTML)` etc. It's not that bad but we already have some existing code that relies on `editor.canvas` being able to call `evaluate`. I haven't found a way to properly mitigate that in gutenberg, but perhaps we can try that in core? But I guess consistency also matters here 😅 .

WunderBart replied:
Just pushed migrated tests that use the `frameLocator` API instead of `frame` in https://github.com/WordPress/gutenberg/pull/54911. It required a bit of juggling, but ultimately it wasn't that bad (assuming, of course, that the tests are going to pass! 😅). Let me know what you think! 🙌 

Re: evaluate from frameLocator

It looks like [the recommended approach](https://github.com/microsoft/playwright/issues/22924#issuecomment-1542498390) is to call 

```js
editor.canvas.locator( ':root' ).evaluate( /**/ );
```

WunderBart replied:
It's a bit more work than I anticipated (of course 🙄), but it's going well. The worst part to figure out so far was drag&drop via evaluated `DataTransfer`. TIL that the DataTransfer can't be transferred from one frame to another (e.g., from the main page to the editor canvas). It needs to be instantiated and dropped in the same frame. [Here's the commit that addressed that](https://github.com/WordPress/gutenberg/pull/54911/commits/51a3a7515598add637d705106782af994e398447).

WunderBart replied:
This is now ready to review! https://github.com/WordPress/gutenberg/pull/54911 cc @kevin940726 @swissspidy

WunderBart replied:
...and it's merged! 🎉

Mamaduka committed:
Update empty and restore trash tests

Mamaduka committed:
Try using same PW version as Gutenberg and wp-scripts

swissspidy commented on package.json:
```diff
@@ -25,7 +25,7 @@
 	"devDependencies": {
 		"@lodder/grunt-postcss": "^3.1.1",
-	  	"@playwright/test": "1.38.1",
+	  	"@playwright/test": "1.32.0",
```
Note to self: we should update the version in Gutenberg as it's wholly out of date

Mamaduka replied:
I've got a PR for it, but there's a regression in a recent version which breaks a few tests. Ref: https://github.com/WordPress/gutenberg/pull/54051#issuecomment-1698886612.

swissspidy committed:
Install deps again after git checkout

swissspidy committed:
Remove commented out config

swissspidy commented on tests/e2e/specs/hello.test.js:
```diff
@@ -1,11 +1,13 @@
-			'//h2[contains(text(), "Welcome to WordPress!")]'
-		);
-		expect( nodes.length ).not.toEqual( 0 );
+test.describe( 'Site editor title', () => {
```
```suggestion
test.describe( 'Hello World', () => {
```

swissspidy committed:
Fix test title

swissspidy committed:
Use env var for prefix instead of argument

swissspidy committed:
Fix app passwords tests

swissspidy committed:
Debug: use different reporter on CI

Mamaduka committed:
Update app password tests

swissspidy committed:
Remove leftovers

swissspidy committed:
Move order

swissspidy committed:
Make Gutenberg test more robust

swissspidy committed:
Make comparison script more robust

Mamaduka commented on tests/e2e/specs/gutenberg-plugin.test.js:
```diff
@@ -1,24 +1,46 @@
+		if ( ! pluginsMap.gutenberg ) {
+			await requestUtils.rest( {
+				method: 'POST',
+				path: 'wp/v2/plugins?slug=gutenberg',
```
I think `slug` should be part of the `body` since this is a `POST` request.

swissspidy replied:
I thought so too originally, but the controller expects slug as a query param.

Mamaduka replied:
That is a bit confusing, especially after reading the PHPUnit tests - https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/rest-api/rest-plugins-controller.php#L384-L385.

swissspidy replied:
Hmm weird, then perhaps I did something wrong. But the tests passes this way, the request works with the query param.

Mamaduka replied:
Let's double-check this before finalizing things.

Referencing the endpoint announcement post for the future: https://make.wordpress.org/core/2020/07/16/new-and-modified-rest-api-endpoints-in-wordpress-5-5/

Mamaduka replied:
It looks like `slug` works as a query argument and as body data.

swissspidy committed:
Further improve comparison script

swissspidy committed:
Undo reporter change

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Remove `PUPPETEER_SKIP_DOWNLOAD`

Browsers are only installed when running tests

swissspidy committed:
Update readmes

Mamaduka committed:
Make Gutenberg plugin tests more resilient

Mamaduka commented on tests/e2e/specs/gutenberg-plugin.test.js:
```diff
@@ -1,26 +1,48 @@
 
-	afterAll( async () => {
-		await uninstallPlugin( 'gutenberg' );
+		// Refetch installed plugin details. It avoids stale values when the test installs the plugin.
+		await requestUtils.getPluginsMap( /* forceRefetch */ true );
```
This prevents tests from failing with `Error: The plugin "gutenberg" isn't installed` when installing the plugin.

swissspidy replied:
Right, because `deactivateMap` calls `getPluginsMap()`. Makes sense 👍

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

desrosj commented on .gitignore:
```diff
@@ -100,4 +100,4 @@ wp-tests-config.php
 
 # Visual regression test diffs
-tests/visual-regression/specs/__image_snapshots__
+tests/visual-regression/specs/__snapshots__
```
If someone has run the tests with the previous setup, they'll have the `__image_snapshots__` left over. Don't feel strongly, but should we leave both of these?

swissspidy replied:
Good point. Looks like Playwright also has a `snapshotPathTemplate` option that we could use to keep the existing `__image_snapshots__` location. Though not sure what happens if you run the new tests with old existing screenshots 🤔 That might also be confusing.

So yeah I suppose we could keep both for the time being.

desrosj replied:
I think it's likely that very few contributors are running these locally. Seems we could probably leave this as it is currently in your changes. Maybe a quick post on Make after this lands to document the change and suggest people can delete the previous snapshot directory.

swissspidy replied:
Sounds good 👍

desrosj commented on .github/workflows/performance.yml:
```diff
@@ -119,6 +121,9 @@ jobs:
       - name: Install npm dependencies
         run: npm ci
 
```
Not necessarily a blocker, but I wonder if we could cache these binaries with actions/cache to reduce the time to fetch these with identical configurations?

Mamaduka replied:
Caching browser binaries is discouraged by Playwright.

Refs:

* https://playwright.dev/docs/ci#caching-browsers
* https://github.com/microsoft/playwright/pull/23814

swissspidy replied:
Not sure if it's really worth it. It could even be slower than just downloading the binaries anew every time.
 
Installing the browsers takes [40 seconds](https://github.com/WordPress/wordpress-develop/actions/runs/6385308130/job/17329768825?pr=5333) right now, which is a small part of the total duration time.

desrosj replied:
Works for me!

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

# Conflicts:
#	package-lock.json
#	package.json

swissspidy committed:
Extend base PW config from wp-scripts

swissspidy committed:
Fix config

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy commented on package.json:
```diff
@@ -25,10 +25,12 @@
 	],
 	"devDependencies": {
 		"@lodder/grunt-postcss": "^3.1.1",
+	  	"@playwright/test": "1.32.0",
```
```suggestion
		"@playwright/test": "1.32.0",
```

swissspidy committed:
Fix indentation

swissspidy committed:
use `require.resolve`

hellofromtonya committed:
Options, Meta APIs: Check setting group exists before search in unregister_setting().

Checks if the given `$option_group` exists before searching for the `$option_name`. Sets the `$pos` to `false`, as `array_search()` returns `false` if the option name (needle) does not exist.

This changeset fixes 2 different PHP Warning|Notice scenarios:
1. When the global `$new_allowed_options` is `null`, fixes raising `Trying to access array offset on value of type null` PHP Notice (PHP 7.4) | Warning (on PHP 8).

2. When the global `$new_allowed_options` is an `array` and the setting group key does not exist, fixes raising "Undefined index: unknown_setting_group" PHP Notice (PHP 7) | Warning (on PHP 8).

For both scenarios, the `array_search()` is skipped and the `$pos` is set to a default of `false`, i.e. which is the value returned when `array_search()` is unsuccessful.

Props xknown, hellofromTonya, nicolefurlan, oglekler, SergeyBiryukov, shailu25.
Fixes #57674.

git-svn-id: https://develop.svn.wordpress.org/trunk@56817 602fd350-edb4-49c9-b593-d223f7449a82

ockham committed:
Patterns: Inject `theme` attribute into Template Part blocks.

[56805] introduced a regression: The `theme` attribute was no longer injected into Template Part blocks inside of patterns. This caused errors on the frontend, where instead of a given template part, an error message such as `Template part has been deleted or is unavailable: header` was seen.

This changeset rectifies that problem, and adds unit test coverage to guard against future regressions.

Follow-up to [56805].
Props scruffian, gziolo.
Fixes #59583.

git-svn-id: https://develop.svn.wordpress.org/trunk@56818 602fd350-edb4-49c9-b593-d223f7449a82

spacedmonkey committed:
REST API: Fix issue with Template and Template Part Revision/Autosave REST API controllers. 

The Template and Template Part REST API controllers have unique characteristics compared to other post type REST API controllers. They do not rely on integer IDs to reference objects; instead, they use a combination of the theme name and slug of the template, like 'twentytwentyfour//home.' Consequently, when the post types template and template part were introduced in [52062], it led to the registration of REST API endpoints for autosaves and revisions with invalid URL structures.

In this commit, we introduce new functionality to enable custom autosave and revisions endpoints to be registered at the post type level. Similar to the 'rest_controller_class' parameter, developers can now define 'revisions_rest_controller' and 'autosave_rest_controller.' This empowers developers to create custom controllers for these functionalities. Additionally, we introduce a 'late_route_registration' parameter, which proves helpful when dealing with custom URL patterns and regex pattern matching issues.
This commit registers new classes for template and template part autosave and revisions controllers, differentiating them from standard controllers in the following ways:
* The response shape now matches that of the template controller.
* Permission checks align with the template controller.
* A custom URL pattern is introduced to support slug-based identification of templates.

Furthermore, we've updated the utility function '_build_block_template_result_from_post' to support passing revision post objects. This enhancement ensures compatibility with the custom revisions controller.

Props spacedmonkey, revgeorge, andraganescu, hellofromTonya, antonvlasenko, kadamwhite, ironprogrammer, costdev, mukesh27, timothyblynjacobs, adamsilverstein. 
Fixes 56922.

git-svn-id: https://develop.svn.wordpress.org/trunk@56819 602fd350-edb4-49c9-b593-d223f7449a82

hellofromtonya committed:
Update/Install: Deactivate Gutenberg plugin version older than 16.5.

This commit changes the Gutenberg minimum compatible version number from 14.1 (introduced in [54790]) to 16.5. For versions older than 16.5, the plugin will deactivate when upgrading WordPress to 6.4-beta3 or newer.

Changes are done within Core's `_upgrade_core_deactivate_incompatible_plugins()` which is invoked during WordPress' upgrade process.

Follow-up to [54790].

Props hellofromTonya, spacedmonkey.
Fixes #59584.

git-svn-id: https://develop.svn.wordpress.org/trunk@56820 602fd350-edb4-49c9-b593-d223f7449a82

karmatosed committed:
WordPress 6.4 Beta 3.


git-svn-id: https://develop.svn.wordpress.org/trunk@56821 602fd350-edb4-49c9-b593-d223f7449a82

karmatosed committed:
Post WordPress 6.4 Beta 3 version bump.


git-svn-id: https://develop.svn.wordpress.org/trunk@56822 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Correct variable typo.

This was causing the version of WordPress being tested to be listed as `latest` even when that is not the case.

Props davidbaumwald.
See #58977.

git-svn-id: https://develop.svn.wordpress.org/trunk@56823 602fd350-edb4-49c9-b593-d223f7449a82

costdev committed:
Plugins: Fix broken `sprintf()` call in plugins list table.

In [56599], a `sprintf()` call was modified which resulted in an insufficient number of arguments.
This caused a Fatal Error when an incompatible plugin notice was displayed.

This fixes the `sprintf()` call.

Follow-up to [56599].

Props petitphp, TobiasBg, sabernhardt, mukesh27.
Fixes #59590. See #57791.

git-svn-id: https://develop.svn.wordpress.org/trunk@56824 602fd350-edb4-49c9-b593-d223f7449a82

costdev committed:
Docs: Use US spelling and correct a typing mistake.

This changes two inline comments and a docblock so that they use US spelling as advised by the Core Handbook's Best Practices. A typing mistake is also corrected.

Reference:
- [https://make.wordpress.org/core/handbook/best-practices/spelling/ Core Handbook - Best Practices - Spelling].

Follow-up to [18632], [38120], [44954].

Props kebbet, mukesh27.
See #58833.

git-svn-id: https://develop.svn.wordpress.org/trunk@56825 602fd350-edb4-49c9-b593-d223f7449a82

SergeyBiryukov committed:
Help/About: Ensure that focus outline on the Credits screen is not cut off.

Props ivanzhuck, oglekler, wildworks, dhrumilk, audrasjb, tejadev, ankit-k-gupta, sumitbagthariya16, mukesh27, marybaum.
Fixes #59033.

git-svn-id: https://develop.svn.wordpress.org/trunk@56826 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Don’t send Slack notification for workflow retries.

[56780] changed Slack failure notifications to not send a failure notification during the first run of a workflow. Because workflows automatically restart once, a failure during the first run can be ignored.

This adjusts the workflow to also not send a “fixed” notification when the second run of a workflow succeeds after a failure. Because the original failure is no longer reported, these notifications are unnecessary.

See #58867.

git-svn-id: https://develop.svn.wordpress.org/trunk@56827 602fd350-edb4-49c9-b593-d223f7449a82

felixarntz committed:
Editor: Add further test coverage for `wp_render_elements_support()`.

Props dmsnell, aaronrobertshaw.
Fixes #59578.


git-svn-id: https://develop.svn.wordpress.org/trunk@56828 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Increase the number of retries when restarting a workflow.

This increases the number of times to retry restarting a workflow from 2 to 10. Retries use exponential backoff to space out retries.

In most cases, only 2 retries was sufficient. However, there are occasionally scenarios where the original workflow is still running and cannot be restarted, resulting in an error. This typically happens during periods of ongoing service degradation, or workflows with a significant number of jobs (GitHub Actions is sometimes very slow to mark these as finished).

See #58867.

git-svn-id: https://develop.svn.wordpress.org/trunk@56829 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Increase the timeout for the failed workflow.

Following [56829], the previous `timeout-minutes` value of `5` is insufficient when approaching 10 retries.

See #58867.

git-svn-id: https://develop.svn.wordpress.org/trunk@56830 602fd350-edb4-49c9-b593-d223f7449a82

desrosj committed:
Build/Test Tools: Remove hardcoded PHPUnit config files.

The configuration file passed to the callable workflow contains the correct one to use.

See #58955.

git-svn-id: https://develop.svn.wordpress.org/trunk@56831 602fd350-edb4-49c9-b593-d223f7449a82

Mamaduka reviewed (approved):
The e2e test migrations look good to me ✅

SergeyBiryukov committed:
Twenty Nineteen: Add margins to editor iframe content.

At screen widths of at least 768px (tablet size), Twenty Nineteen should have a `max-width` and side margins for content in the post editor.

This commit applies that style to the new `.block-editor-iframe__body` class, resolving an issue where full-width blocks had a horizontal scrollbar in the editor iframe.

Props sabernhardt, smit08, nidhidhandhukiya, mukesh27, darshitrajyaguru97, oglekler, poena, shailu25, nicolefurlan.
Fixes #59449.

git-svn-id: https://develop.svn.wordpress.org/trunk@56832 602fd350-edb4-49c9-b593-d223f7449a82

kevin940726 commented on tests/e2e/specs/edit-posts.test.js:
```diff
@@ -1,137 +1,135 @@
-			`//a[contains(@class, "row-title")][contains(text(), "${ title }")]`
-		);
-		await editLink.focus();
+		// // Focus on the post title link.
```
😅 
```suggestion
		// Focus on the post title link.
```

kevin940726 reviewed (approved):
LGTM 👍 I didn't spot any issues but could also use some more pairs of eyes 💯.

swissspidy committed:
Fix double comment

Co-authored-by: Kai Hao <kevin830726@gmail.com>

swissspidy committed:
Mention UI mode in readme

desrosj reviewed (approved):
The PR seems muddied up a bit from a rebase or merge. But I think this is good to give a go!

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Revert r56198

Partially reverts https://core.trac.wordpress.org/changeset/56198/

swissspidy committed:
Apply r56660

See https://core.trac.wordpress.org/changeset/56660/

swissspidy committed:
Fix env vars

swissspidy committed:
Merge branch 'trunk' into add/pw-performance-tests

swissspidy committed:
Merge branch 'WordPress:trunk' into add/pw-performance-tests

adamsilverstein reviewed (commented):
Overall looks good, I wasn't able to get tests running due to a local issue.

swissspidy commented:
Committed in https://core.trac.wordpress.org/changeset/56926 🎉 

Thanks y'all!
